lib/sync/ephy-history-manager.c

```diff
@@ -0,0 +1,563 @@
+
+G_DEFINE_TYPE_WITH_CODE (EphyHistoryManager, ephy_history_manager, G_TYPE_OBJECT,
+                         G_IMPLEMENT_INTERFACE (EPHY_TYPE_SYNCHRONIZABLE_MANAGER,
+                                                ephy_synchronizable_manager_iface_init))
+
+enum {
+  PROP_0,
+  PROP_HISTORY_SERVICE,
+  LAST_PROP
+};
+
+static GParamSpec *obj_properties[LAST_PROP];
+
+typedef struct {
+  EphyHistoryManager                     *manager;
+  gboolean                                is_initial;
+  GSList                                 *remotes_deleted;
+  GSList                                 *remotes_updated;
+  EphySynchronizableManagerMergeCallback  callback;
+  gpointer                                user_data;
+} MergeHistoryAsyncData;
+
+static MergeHistoryAsyncData *
+merge_history_async_data_new (EphyHistoryManager                     *manager,
+                              gboolean                                is_initial,
+                              GSList                                 *remotes_deleted,
+                              GSList                                 *remotes_updated,
+                              EphySynchronizableManagerMergeCallback  callback,
+                              gpointer                                user_data)
+{
+  MergeHistoryAsyncData *data;
+
+  data = g_slice_new (MergeHistoryAsyncData);
+  data->manager = g_object_ref (manager);
+  data->is_initial = is_initial;
+  data->remotes_deleted = remotes_deleted;
+  data->remotes_updated = remotes_updated;
+  data->callback = callback;
+  data->user_data = user_data;
+
+  return data;
+}
+
+static void
+merge_history_async_data_free (MergeHistoryAsyncData *data)
+{
+  g_assert (data);
+
+  g_object_unref (data->manager);
+  g_slice_free (MergeHistoryAsyncData, data);
+}
+
+static void
+url_visited_cb (EphyHistoryService *service,
+                EphyHistoryURL     *url,
+                EphyHistoryManager *self)
+{
+  EphyHistoryRecord *record;
+
+  record = ephy_history_record_new (url->sync_id, url->title, url->url, url->last_visit_time);
+  g_signal_emit_by_name (self, "synchronizable-modified", record, TRUE);
+  g_object_unref (record);
+}
+
+static void
+url_deleted_cb (EphyHistoryService *service,
+                EphyHistoryURL     *url,
+                EphyHistoryManager *self)
+{
+  EphyHistoryRecord *record;
+
+  record = ephy_history_record_new (url->sync_id, url->title, url->url, url->last_visit_time);
+  g_signal_emit_by_name (self, "synchronizable-deleted", record);
+  g_object_unref (record);
+}
+
+static void
+ephy_history_manager_set_property (GObject      *object,
+                                   guint         prop_id,
+                                   const GValue *value,
+                                   GParamSpec   *pspec)
+{
+  EphyHistoryManager *self = EPHY_HISTORY_MANAGER (object);
+
+  switch (prop_id) {
+    case PROP_HISTORY_SERVICE:
+      g_clear_object (&self->service);
```

Comment:
Just use g_object_unref here. There's no point in setting it to NULL immediately before assigning to it.